control/plugin/execution.go

```diff
@@ -125,222 +58,72 @@ func NewExecutablePlugin(a Arg, path string) (*ExecutablePlugin, error) {
-				p.Kill()
-				break
-			}
-			log.Debug("timeout flag ignored because of response")
-
-		case pluginKilled: // plugin killed signal received
-			log.Error("plugin kill signal received")
-			// We check a few scenarios and return based on how things worked out to this point
-			// 1) If a bad response was received we return signalling this with an error (fail)
-			if errResponse != nil {
-				log.Error("returning with error (bad response)")
-				return nil, *errResponse
-			}
-			// 2) If a timeout occurred we return that as error (fail)
-			if timeoutFlag {
-				log.Error("returning with error (timeout)")
-				return nil, errors.New("timeout waiting for response")
-			}
-			// 3) If a good response was returned we return that with no error (success)
-			if response != nil {
-				log.Error("returning with response (after wait for kill)")
-				return response, nil
-			}
-			// 4) otherwise we return no response and an error that no response was received (fail)
-			log.Error("returning with error (killed without response)")
-			// The kill could have been without error so we check if ExitError was returned and return
-			// our own if not.
-			if *w.Error != nil {
-				return nil, *w.Error
+// Run executes the plugin and waits for a response, or times out.
+func (e *ExecutablePlugin) Run() (Response, error) {
+	var (
+		respReceived bool
+		resp         Response
+		err          error
+	)
+
+	doneChan := make(chan struct{})
+	stdOutScanner := bufio.NewScanner(e.stdout)
+
+	// Start the command and begin reading its output.
+	e.cmd.Start()
+	e.captureStderr()
+	go func() {
+		for stdOutScanner.Scan() {
+			// The first chunk from the scanner is the plugin's response to the
+			// handshake.  Once we've received that, we can begin to forward
+			// logs on to snapd's log.
+			if !respReceived {
+				respBytes := stdOutScanner.Bytes()
+				err = json.Unmarshal(respBytes, &resp)
+				respReceived = true
+				close(doneChan)
 			} else {
-				return nil, errors.New("plugin died without sending response")
+				execLogger.Info(fmt.Sprintf("%s plugin: %s",
+					path.Base(e.cmd.Path), stdOutScanner.Text()))
 			}
-
-		case pluginResponseOk: // plugin response (valid) signal received
-			log.Debug("plugin response (ok) received")
-			// If in daemon mode we can return now (succes) since the plugin will continue to run
-			// if not we let the loop continue (to wait for kill)
-			response = w.Response
-			return response, nil
-
-		case pluginResponseBad: // plugin response (invalid) signal received
-			log.Error("plugin response (bad) received")
-			// A bad response is end of game in all scerarios and indictive of an unhealthy or unsupported plugin
-			// We save the response bad error var (for handling later on plugin kill)
-			errResponse = w.Error
 		}
+	}()
+
+	// Wait until:
+	//   a) We receive a signal that the plugin has responded
+	// OR
+	//   b) The timeout expires
+	select {
+	case <-doneChan:
+	case <-time.After(time.Millisecond * 1500):
```

Comment:
This should be tunable. On a busy system or a small device it could be much longer.